File: alura/java-oo/banco2/Conta.java
public class Conta {
  private double saldo;
  private String titular;
  private int agencia;
  private int numero;

  public void saca(double valor) throws Exception {
    if(saldo >= valor) {
      this.saldo -= valor;
    } else {
      throw new SaldoInsuficienteException(saldo);
    }
  }

  public void deposita(double valor) {
    if(valor > 0) {
      this.saldo += valor;
    } else {
      throw new ValorInvalidoException(valor);
    }
  }

  public String getTitular() {
    return titular;
  }

  public void setTitular(String titular) {
    this.titular = titular;
  }

  public double getSaldo() {
    return this.saldo;
  }

  //abstract public void atualiza(double taxaSelic);
}

File: alura/c-sharp/Conta.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Caelum.CaixaEletronico.Usuarios;

namespace Caelum.CaixaEletronico.Contas
{
  abstract class Conta
  {
    public int Numero { get; set; }

    public double Saldo { get; protected set; }

    public Cliente Titular { get; set; }

    public int Tipo { get; set; }

    public void Saca(double valor)
    {
      if(valor < 0)
      {
        throw new ArgumentException();
      }

      if(this.Saldo >= valor)
      {
        this.Saldo -= valor;
      }
      else
      {
        throw new SaldoInsuficienteException();
      }
    }

    public void Deposita(double valor)
    {
      this.Saldo += valor;
    }

    public void Transfere(double valor, Conta destino)
    {
      this.Saca(valor);
      destino.Deposita(valor);
    }
  }
}
